File: game_2048.py
import numpy as np
import copy

class Game2048:
    def __init__(self):
        self.board = np.zeros((4, 4), dtype=int)
        self.score = 0
        self.is_player_turn = True  # True表示移动方向的玩家，False表示放置数字的玩家
        self.step = 0
        self.add_new_tile()
        self.add_new_tile()
        
    def add_new_tile(self):
        empty_cells = list(zip(*np.where(self.board == 0)))
        if empty_cells:
            x, y = empty_cells[np.random.randint(len(empty_cells))]
            self.board[x, y] = 2 if np.random.random() < 0.9 else 4
            
    def get_state(self):
        return self.board.copy()

    def get_step(self):
        return self.step
    
    def get_score(self):
        return self.score
        
    def is_game_over(self):
        if np.any(self.board == 0):
            return False
        for i in range(4):
            for j in range(4):
                if i < 3 and self.board[i, j] == self.board[i + 1, j]:
                    return False
                if j < 3 and self.board[i, j] == self.board[i, j + 1]:
                    return False
        return True
    
    # 将(pos(x, y), value)转化为id线性动作空间
    def place_tile_id(self, vid):
        id = vid % 16
        value = 2 if id < 16 else 4
        """放置数字玩家的动作"""
        if self.board[id // 4, id % 4] == 0:
            self.board[id // 4, id % 4] = value
            self.is_player_turn = True
            return True
        return False
    
    def move(self, direction):
        """移动方向玩家的动作"""
        if not self.is_player_turn:
            return False
            
        # 0: up, 1: right, 2: down, 3: left
        original_board = self.board.copy()
        original_score = self.score
        
        # 统一转换为向左移动的情况
        if direction == 0:  # up
            self.board = np.rot90(self.board, k=1)
        elif direction == 1:  # right
            self.board = np.rot90(self.board, k=2)
        elif direction == 2:  # down
            self.board = np.rot90(self.board, k=3)
        
        # 对每一行进行处理
        for i in range(4):
            # 移除零并获取非零数字
            non_zero = self.board[i][self.board[i] != 0]
            
            # 合并相同的数字
            if len(non_zero) >= 2:
                result = []
                j = 0
                while j < len(non_zero):
                    if j + 1 < len(non_zero) and non_zero[j] == non_zero[j+1]:
                        merged = non_zero[j] * 2
                        self.score += merged
                        result.append(merged)
                        j += 2
                    else:
                        result.append(non_zero[j])
                        j += 1
                non_zero = np.array(result)
            
            # 用零填充到长度4
            new_line = np.zeros(4, dtype=int)
            if len(non_zero) > 0:
                new_line[:len(non_zero)] = non_zero
            self.board[i] = new_line
        
        # 旋转回原来的方向
        if direction == 0:  # up
            self.board = np.rot90(self.board, k=3)
        elif direction == 1:  # right
            self.board = np.rot90(self.board, k=2)
        elif direction == 2:  # down
            self.board = np.rot90(self.board, k=1)
            
        # 检查板子是否发生变化
        if not np.array_equal(original_board, self.board):
            self.is_player_turn = False
            self.step += 1
            return True
        else:
            self.score = original_score
            return False
            
    def get_max_tile(self):
        return np.max(self.board)
    
    def get_available_moves(self):
        """获取当前玩家的可用动作"""
        if self.is_player_turn:
            # 移动方向玩家的可用动作
            available_moves = []
            for direction in range(4):
                game_copy = copy.deepcopy(self)
                if game_copy.move(direction):
                    available_moves.append(direction)
            return available_moves
        else:
            # 放置数字玩家的可用动作
            empty_cells = self.get_empty_cells()
            available_moves = []
            for cell in empty_cells:
                id = cell[0] * 4 + cell[1]
                available_moves.append(id)
            for cell in empty_cells:
                id = cell[0] * 4 + cell[1] + 16
                available_moves.append(id)
            return available_moves
    
    def get_empty_cells(self):
        return list(zip(*np.where(self.board == 0)))
    
    def clone(self):
        return copy.deepcopy(self)
    
    def __str__(self):
        return str(self.board) 

File: alpha_zero_golang/game_2048.go
package main

import (
	"fmt"
	"math/rand"
)

// Game2048 represents the 2048 game state
type Game2048 struct {
	board        [4][4]int
	score        int
	isPlayerTurn bool // true表示移动方向的玩家，false表示放置数字的玩家
	step         int
}

// NewGame2048 creates and initializes a new game
func NewGame2048() *Game2048 {
	game := &Game2048{
		board:        [4][4]int{},
		score:        0,
		isPlayerTurn: true,
		step:         0,
	}
	game.addNewTile()
	game.addNewTile()
	return game
}

// Clone creates a deep copy of the game
func (g *Game2048) Clone() *Game2048 {
	newGame := &Game2048{
		board:        [4][4]int{},
		score:        g.score,
		isPlayerTurn: g.isPlayerTurn,
		step:         g.step,
	}
	copy(newGame.board[:], g.board[:])
	return newGame
}

// addNewTile adds a new tile (2 or 4) to a random empty cell
func (g *Game2048) addNewTile() {
	emptyCells := g.getEmptyCells()
	if len(emptyCells) == 0 {
		return
	}

	cell := emptyCells[rand.Intn(len(emptyCells))]
	value := 2
	if rand.Float64() > 0.9 {
		value = 4
	}
	g.board[cell[0]][cell[1]] = value
}

// getEmptyCells returns a list of empty cell coordinates
func (g *Game2048) getEmptyCells() [][2]int {
	emptyCells := make([][2]int, 0, 16) // 预分配16个空间，避免growslice
	for i := range g.board {
		for j := range g.board[i] {
			if g.board[i][j] == 0 {
				emptyCells = append(emptyCells, [2]int{i, j})
			}
		}
	}
	return emptyCells
}

// Move performs a move in the specified direction
// direction: 0=up, 1=right, 2=down, 3=left
func (g *Game2048) Move(direction int) bool {
	if !g.isPlayerTurn {
		return false
	}

	var moved bool
	originalScore := g.score

	switch direction {
	case 0: // 上移
		moved = g.moveUp()
	case 1: // 右移
		moved = g.moveRight()
	case 2: // 下移
		moved = g.moveDown()
	case 3: // 左移
		moved = g.moveLeft()
	}

	if !moved {
		g.score = originalScore
		return false
	}

	g.isPlayerTurn = false
	g.step++
	return true
}

// moveLeft moves all tiles to the left and merges them
func (g *Game2048) moveLeft() bool {
	moved := false
	for i := 0; i < 4; i++ {
		// 压缩行，移除空格
		line := make([]int, 0, 4)
		for j := 0; j < 4; j++ {
			if g.board[i][j] != 0 {
				line = append(line, g.board[i][j])
			}
		}

		// 如果这一行没有数字，跳过
		if len(line) == 0 {
			continue
		}

		// 合并相同的数字
		for j := 0; j < len(line)-1; j++ {
			if line[j] == line[j+1] {
				line[j] *= 2
				g.score += line[j]
				// 将后面的数字前移
				copy(line[j+1:], line[j+2:])
				line = line[:len(line)-1]
				moved = true
			}
		}

		// 检查是否发生了移动
		for j := 0; j < 4; j++ {
			newValue := 0
			if j < len(line) {
				newValue = line[j]
			}
			if g.board[i][j] != newValue {
				moved = true
			}
			g.board[i][j] = newValue
		}
	}
	return moved
}

// moveRight moves all tiles to the right and merges them
func (g *Game2048) moveRight() bool {
	moved := false
	for i := 0; i < 4; i++ {
		// 压缩行，移除空格
		line := make([]int, 0, 4)
		for j := 3; j >= 0; j-- {
			if g.board[i][j] != 0 {
				line = append(line, g.board[i][j])
			}
		}

		// 如果这一行没有数字，跳过
		if len(line) == 0 {
			continue
		}

		// 合并相同的数字
		for j := 0; j < len(line)-1; j++ {
			if line[j] == line[j+1] {
				line[j] *= 2
				g.score += line[j]
				// 将后面的数字前移
				copy(line[j+1:], line[j+2:])
				line = line[:len(line)-1]
				moved = true
			}
		}

		// 检查是否发生了移动
		for j := 0; j < 4; j++ {
			newValue := 0
			if j < len(line) {
				newValue = line[j]
			}
			if g.board[i][3-j] != newValue {
				moved = true
			}
			g.board[i][3-j] = newValue
		}
	}
	return moved
}

// moveUp moves all tiles up and merges them
func (g *Game2048) moveUp() bool {
	moved := false
	for j := 0; j < 4; j++ {
		// 压缩列，移除空格
		line := make([]int, 0, 4)
		for i := 0; i < 4; i++ {
			if g.board[i][j] != 0 {
				line = append(line, g.board[i][j])
			}
		}

		// 如果这一列没有数字，跳过
		if len(line) == 0 {
			continue
		}

		// 合并相同的数字
		for i := 0; i < len(line)-1; i++ {
			if line[i] == line[i+1] {
				line[i] *= 2
				g.score += line[i]
				// 将后面的数字前移
				copy(line[i+1:], line[i+2:])
				line = line[:len(line)-1]
				moved = true
			}
		}

		// 检查是否发生了移动
		for i := 0; i < 4; i++ {
			newValue := 0
			if i < len(line) {
				newValue = line[i]
			}
			if g.board[i][j] != newValue {
				moved = true
			}
			g.board[i][j] = newValue
		}
	}
	return moved
}

// moveDown moves all tiles down and merges them
func (g *Game2048) moveDown() bool {
	moved := false
	for j := 0; j < 4; j++ {
		// 压缩列，移除空格
		line := make([]int, 0, 4)
		for i := 3; i >= 0; i-- {
			if g.board[i][j] != 0 {
				line = append(line, g.board[i][j])
			}
		}

		// 如果这一列没有数字，跳过
		if len(line) == 0 {
			continue
		}

		// 合并相同的数字
		for i := 0; i < len(line)-1; i++ {
			if line[i] == line[i+1] {
				line[i] *= 2
				g.score += line[i]
				// 将后面的数字前移
				copy(line[i+1:], line[i+2:])
				line = line[:len(line)-1]
				moved = true
			}
		}

		// 检查是否发生了移动
		for i := 0; i < 4; i++ {
			newValue := 0
			if i < len(line) {
				newValue = line[i]
			}
			if g.board[3-i][j] != newValue {
				moved = true
			}
			g.board[3-i][j] = newValue
		}
	}
	return moved
}

// IsGameOver checks if the game is over
func (g *Game2048) IsGameOver() bool {
	// Check for empty cells
	for i := 0; i < 4; i++ {
		for j := 0; j < 4; j++ {
			if g.board[i][j] == 0 {
				return false
			}
		}
	}

	// Check for possible merges
	for i := 0; i < 4; i++ {
		for j := 0; j < 4; j++ {
			if i < 3 && g.board[i][j] == g.board[i+1][j] {
				return false
			}
			if j < 3 && g.board[i][j] == g.board[i][j+1] {
				return false
			}
		}
	}
	return true
}

// GetAvailableMoves returns a list of valid moves
func (g *Game2048) GetAvailableMoves() []int {
	if g.isPlayerTurn {
		// 移动方向玩家的可用动作
		var moves []int
		for direction := 0; direction < 4; direction++ {
			gameCopy := g.Clone()
			if gameCopy.Move(direction) {
				moves = append(moves, direction)
			}
		}
		return moves
	} else {
		// 放置数字玩家的可用动作
		emptyCells := g.getEmptyCells()
		moves := make([]int, 0, len(emptyCells)*2) // 预分配空间，每个空格可以放2或4
		for _, cell := range emptyCells {
			id := cell[0]*4 + cell[1]
			moves = append(moves, id) // 放置2的动作
		}
		for _, cell := range emptyCells {
			id := cell[0]*4 + cell[1]
			moves = append(moves, id+16) // 放置4的动作
		}
		return moves
	}
}

// GetScore returns the current score
func (g *Game2048) GetScore() int {
	return g.score
}

// GetMaxTile returns the highest tile value on the board
func (g *Game2048) GetMaxTile() int {
	maxTile := 0
	for i := 0; i < 4; i++ {
		for j := 0; j < 4; j++ {
			if g.board[i][j] > maxTile {
				maxTile = g.board[i][j]
			}
		}
	}
	return maxTile
}

// GetBoard returns the current board state as [4][4]int
func (g *Game2048) GetBoard() [4][4]int {
	var board [4][4]int
	copy(board[:], g.board[:])
	return board
}

// String returns a string representation of the board
func (g *Game2048) String() string {
	var result string
	for i := range g.board {
		for j := range g.board[i] {
			result += fmt.Sprintf("%4d ", g.board[i][j])
		}
		result += "\n"
	}
	return result
}

// PlaceTileID places a new tile at the specified position
// vid is the action ID, where vid % 16 determines the position and vid >= 16 determines if it's a 4
func (g *Game2048) PlaceTileID(vid int) bool {
	if g.isPlayerTurn {
		return false
	}

	id := vid % 16
	value := 2
	if vid >= 16 {
		value = 4
	}

	row, col := id/4, id%4
	if g.board[row][col] != 0 {
		return false
	}

	g.board[row][col] = value
	g.isPlayerTurn = true
	return true
}

// GetStep returns the current step number
func (g *Game2048) GetStep() int {
	return g.step
}

// IsPlayerTurn returns whether it's the moving player's turn
func (g *Game2048) IsPlayerTurn() bool {
	return g.isPlayerTurn
}
